src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -183,15 +182,31 @@ void ICollection.CopyTo(Array array, int index)
         // Adds item to the tail of the queue.
         public void Enqueue(T item)
         {
-            if (_size == _array.Length)
+            int tail = _tail;
+            int size = _size;
+            T[] array = _array;
+
+            if ((uint)size < (uint)array.Length && (uint)tail < (uint)array.Length)
             {
-                int newcapacity = (int)((long)_array.Length * (long)GrowFactor / 100);
-                if (newcapacity < _array.Length + MinimumGrow)
-                {
-                    newcapacity = _array.Length + MinimumGrow;
-                }
-                SetCapacity(newcapacity);
+                array[tail] = item;
+                MoveNext(ref tail, array);
+                _tail = tail;
+                _size = size + 1;
+                _version++;
             }
+            else
+            {
+                EnqueueWithResize(item);
```

Comment:
What's the benefit to separating this out into a non-inlineable method? Is Enqueue inlined with this separated out?

Reply:
> What's the benefit ...

When `Enqueue` was inlined, it was a hot/cold-path-split. But this doesn't apply anymore.

> Is Enqueue inlined with this separated out?

No. At least not now, as `AggressiveInlining` was removed.

When written as 
```c#
public void Enqueue(T item)
{
    int tail = _tail;
    int size = _size;
    T[] array = _array;

    if ((uint)size < (uint)array.Length && (uint)tail < (uint)array.Length)
    {
        array[tail] = item;
        _tail = MoveNext(ref tail, array);
        _size = size + 1;
        _version++;
    }
    else
    {
        int length = array.Length;
        int newCapaicty = Math.Max(length * 2, length + MinimumGrow);

        SetCapacity(newCapaicty);

        _array[_tail] = item;
        MoveNext(ref _tail);
        _size++;
        _version++;
    }
}
```
(so just manually inlined `EnqueueWithResize`) and with the `_version++` in each branch the asm becomes:
```asm
G_M9246_IG01:
       55                   push     rbp
       4156                 push     r14
       53                   push     rbx
       488D6C2410           lea      rbp, [rsp+10H]
       488BDF               mov      rbx, rdi
       448BF6               mov      r14d, esi

G_M9246_IG02:
       8B7B14               mov      edi, dword ptr [rbx+20]
       8B7318               mov      esi, dword ptr [rbx+24]
       488B4308             mov      rax, gword ptr [rbx+8]
       8B5008               mov      edx, dword ptr [rax+8]
       3BD6                 cmp      edx, esi
       7624                 jbe      SHORT G_M9246_IG05
       3BD7                 cmp      edx, edi
       7620                 jbe      SHORT G_M9246_IG05
       4863CF               movsxd   rcx, edi
       4489748810           mov      dword ptr [rax+4*rcx+16], r14d
       FFC7                 inc      edi
       3BD7                 cmp      edx, edi
       7502                 jne      SHORT G_M9246_IG03
       33FF                 xor      edi, edi

G_M9246_IG03:
       897B14               mov      dword ptr [rbx+20], edi
       FFC6                 inc      esi
       897318               mov      dword ptr [rbx+24], esi
       FF431C               inc      dword ptr [rbx+28]

G_M9246_IG04:
       5B                   pop      rbx
       415E                 pop      r14
       5D                   pop      rbp
       C3                   ret      

G_M9246_IG05:
       8BF2                 mov      esi, edx
       D1E6                 shl      esi, 1
       83C204               add      edx, 4
       3BF2                 cmp      esi, edx
       7D02                 jge      SHORT G_M9246_IG06
       EB02                 jmp      SHORT G_M9246_IG07

G_M9246_IG06:
       8BD6                 mov      edx, esi

G_M9246_IG07:
       488BFB               mov      rdi, rbx
       8BF2                 mov      esi, edx
       E8D3F1FFFF           call     MyQueue`1:SetCapacity(int):this
       488B4308             mov      rax, gword ptr [rbx+8]
       8B7B14               mov      edi, dword ptr [rbx+20]
       8B7008               mov      esi, dword ptr [rax+8]
       3BFE                 cmp      edi, esi
       7323                 jae      SHORT G_M9246_IG10
       4863FF               movsxd   rdi, edi
       448974B810           mov      dword ptr [rax+4*rdi+16], r14d
       488D4314             lea      rax, bword ptr [rbx+20]
       8B38                 mov      edi, dword ptr [rax]
       FFC7                 inc      edi
       3BF7                 cmp      esi, edi
       7502                 jne      SHORT G_M9246_IG08
       33FF                 xor      edi, edi

G_M9246_IG08:
       8938                 mov      dword ptr [rax], edi
       FF4318               inc      dword ptr [rbx+24]
       FF431C               inc      dword ptr [rbx+28]

G_M9246_IG09:
       5B                   pop      rbx
       415E                 pop      r14
       5D                   pop      rbp
       C3                   ret      

G_M9246_IG10:
       E80DD9B678           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3    
```

So there is a "fast-return" at `G_M9246_IG04` (with the same asm as with a separate `EnqueueWithResize`), then followed by the "resize"-part. I think this code is good and an explicitely `EnqueueWithResize` can be omitted.

Note: if `_version++` is at the end of the method, then there is just one single `ret`, which may slow down the fast-path.
Alternatively `_version++` could be moved to the top of the method, so asm becomes
```asm
G_M9246_IG02:
       FF431C               inc      dword ptr [rbx+28]    ; _version++
       8B7B14               mov      edi, dword ptr [rbx+20]
       8B7318               mov      esi, dword ptr [rbx+24]
       488B4308             mov      rax, gword ptr [rbx+8]
       8B5008               mov      edx, dword ptr [rax+8]
       3BD6                 cmp      edx, esi
       7621                 jbe      SHORT G_M9246_IG05
; ...
```

Reply:
> (with the same asm as with a separate `EnqueueWithResize`)

I cheated a bit. In the prolog and epilog there's a difference:
```diff
G_M9243_IG01:
       55                   push     rbp
-      488BEC               mov      rbp, rsp
-      90                   nop
+      4156                 push     r14
+      53                   push     rbx
+      488D6C2410           lea      rbp, [rsp+10H]
+      488BDF               mov      rbx, rdi
+      448BF6               mov      r14d, esi

; ...
G_M9246_IG04:
+      5B                   pop      rbx
+      415E                 pop      r14
       5D                   pop      rbp
       C3                   ret

; ...
```

Don't know if this makes much difference -- one may need to measure it, but I assume it's within noise.

Reply:
This is also followed from `Stack.Push` which splits those out as well
`// Non-inline from Stack.Push to improve its code quality as uncommon path` (added in #26086)

Perhaps I was misunderstanding the purpose behind that in the first place. Even without `AggressiveInlining` the choice may be made to inline it, so separating out the cold-path is beneficial. Any guidance here is definitely appreciated because I might have totally mis-understood this

Reply:
@Wraiyth you haven't missunderstood something. It a usual review-process to question and reason about every change -- even when the change is taken over from an accepted PR, so it's good to re-think about some decisions (made earlier, or made now). That's the case here.